### ISSUE TITLE
(Unexpectedly) Loading "config file" from a different directory than the "runtime directory"

### ISSUE THREAD

**Issue by mrovnanik** [moqui/moqui-framework#78]
I am running Moqui in a Tomcat. I attempted to deploy a new WAR on the same server using connect to a different database (set in MoquiProductionConf.xml). I experienced strange behavior. A new application was started but it loaded config from a different directory (a different application) than it was supposed to load.

Checkout the directories of "runtime directory" and "config file" settings included in the attachment.

![image](https://cloud.githubusercontent.com/assets/14863763/16855080/0dcbdc7c-4a14-11e6-8dd2-aeb4f922da37.png)

**Comment by mrovnanik** [comment]
I suppose this has something to do with ExecutionContextFactoryImpl() which uses system property () if there is one, instead of the properties file. And I suspect the first application running sets the system property to its folder.

**Comment by jonesde** [comment]
You have more than one copy of moqui (i.e. more than one moqui WAR file, expanded to webapp directory) in place? In theory that's possible aside from the system property issue and that would need some changes though I'm not sure quite how best to handle it, but depending on what you are trying to do there are likely better ways to do it. If that is not intentional, just remove one of the moqui webapps. There may be other issues of this as well, I haven't tested that on Tomcat so I don't know if the webapp classpath isolation will handle the static reference to the ExecutionContextFactory in Moqui.java (this is used for things like deserialization to get a reference to the framework).

**Comment by mrovnanik** [comment]
Yes, it was intensional. Quite surprisingly, even for me. The reason was costs. I want to utilize my first server the most. That is the reason why I attempted to run multiple applications on a single Tomcat instance. Each on dedicated database. 

I figured out an idea how the moqui.conf and moqui.runtime are dealt with in runtime. Well, if you yourself are not sure how to deal with this situation (prioritizing system property over moquiInitProperties), it's a strong signal for me to stop thinkink this way.

**Comment by jonesde** [comment]
The problem with the properties file overriding the system properties is that the properties file is always there so the only way to change the runtime configuration would be by changing the properties file.

For the runtime directory location, which is what looks like is getting messed up here (there are other ways of distinguishing just the runtime config file, i.e. put the full path to the file in the properties file instead of relative to the runtime directory), and that will cause issues as well with logging and any local storage you might be doing (i.e. local storage for ElasticSearch instead of a remote instance).

I'll take a look at that, if there is a runtime directory embedded in a WAR file the system property should not override it... or maybe the issue is just with the system property for the config file location, and in the case of a WAR embedded runtime directory it should use the path relative to that instead of the system property.

It may also work to just not set the system property based on the configuration determined through other means... can be nice to have there but isn't really necessary.

As for the bigger picture, there may be better ways to run multiple applications on the same server (assuming different code so the multi-tenant thing is awkward), like using virtual machines or containers, or just different JVM instances. Depending on what the external URLs look like this may be more difficult to setup.

Why do you have multiple applications in the first place?

**Comment by mrovnanik** [comment]
The behavior is like this - it is not the runtime directory location, but it is the configuration file location that gets messed up. I figured it this way:

The first deployed application on my Tomcat works like charm, no problems at all (that is the application 'kinekus-ekonomika').
Then, after I attempt to deploy the second application (the '**test**'), the behavior is as follows:

1. the _MoquiContextListener.contextInitialized_ sets the "moqui.runtime" property to the runtimePath of the embedded application (your comment: // before we init the ECF, see if there is a runtime directory in the webappRealPath, and if so set that as the moqui.runtime System property),
2. from now the moqui.runtime is set to the application directory (which is OK, in my case it's the **test**'s directory),
3. then the _ExecutionContextFactoryImpl_ reads the moqui.runtime property (still OK), but afterwards (on line 121 of ExecutionContextFactoryImpl.groovy) reads the moqui.conf property. And this is the problem, because it's the moqui.conf property set by the previous run, which in my case was the 'kinekus-ekonomika' application WAR.

The unexpected behavior is caused by the fact, that whatever I do to MoquiInit.properties (I set the runtime.conf to a full path, not just a relative one), the runtime just does not take into account, since the system properties do have a priority.

Wouldn't it be a solution to have the code check if it is running as a WAR (?) and in this case use the MoquiInit.properties rather than the system properties? To have the original functionality as it is now and change the functionality when running deployed as a WAR.

You may be right with the options provided, but I am not sure I know how to implement then. Right now I have one application server (with one address and HTTPS set).

**Comment by jonesde** [comment]
The easiest solution is don't set the moqui.conf property in ECFI. The moqui.runtime property gets set for WAR files with an embedded runtime directory in the MoquiContextListener class, but that works for multiple moqui instances because it is set before each one (that could be cleaner, passed internally instead of through system property, but we need to make sure the moqui.runtime system property is set before certain things like Log4J2, BTM, DBs, etc because the default configurations use the moqui.runtime system property to get paths.

This should work fine as long as one instance of moqui fully inits before another one starts, which all servlet containers should do when loading webapps (as opposed to loading in parallel, would cause problems in this scenario).

Could you try the changes in commit #d4de5e3 and see if it works for you now?

**Comment by mrovnanik** [comment]
I just pulled the commit into my fork. I will let you know. Thanks, David.

**Comment by mrovnanik** [comment]
OK, here are the first results:

1. there are two WARs deployed on the Tomcat - test1 and test2 (good result),
2. by checking the respective logs and MoquiActualConf.xmls (in the deployment directories on Tomcat) - each application is connected to its dedicated database (and that is a good result as well),
3. I can open the login screens of both apps (http://srv06.rovnanik.sk/test1/Login, http://srv06.rovnanik.sk/test2/Login),
4. but I cannot log into the app (definitely there is some problem, because the app does not return any issues wih user credentials, it just does not allow entering app) - no message in browser, error logs are included,

So, yes, we moved a step further, but a new issue has shown up. Maybe not an issue, but a configuration error.

![image](https://cloud.githubusercontent.com/assets/14863763/16890731/91b24192-4af1-11e6-8f37-6b41ec8adfd4.png)

![image](https://cloud.githubusercontent.com/assets/14863763/16890739/a0dece24-4af1-11e6-90f1-0448d5065e60.png)

![image](https://cloud.githubusercontent.com/assets/14863763/16890753/c51b7dfa-4af1-11e6-8462-c508d49d8624.png)

![image](https://cloud.githubusercontent.com/assets/14863763/16890839/d158c716-4af2-11e6-8e42-e044135eca84.png)

**Comment by jonesde** [comment]
You'll at least have to give Bitronix different bitronix.tm.serverId values (in the bitronix-default-config.properties file for each app). There may be many more things required... but generally treat them as separate servers included ElasticSearch (often needed) and Hazelcast (only needed for clustered servers, for your case sounds like best not to have in place as it just slows things down) if installed.

**Comment by jonesde** [comment]
On a side note, while it might be nice to make sure the framework works for this sort of thing I doubt it is the best option whatever your situation. These days hardware and virtualization software are so cheap that there is little reason to try to pack small instances into a single deployment, especially with all the complexities involved with that (even in different JVMs on the same host). If you're running on something like AWS Moqui runs fine on a t2.small instance (like the demo.moqui.org server) for just a few dollars a month. If you're running on private hardware it is easy to setup virtualized systems or containers and keep things isolated and clean that way.

More and more I'm even wondering about the utility of the multi-tenant features in Moqui. The whole concept adds all sorts of complexity, and just about everything in the framework has to be tenant-aware. BTW, you can have separate applications for different tenants (set which tenants are allowed to use a screen for the root screen of an application), which you might try. Even then unless you are looking at lots of little simple but different applications (and then what's the point of using something as complex as Moqui...) separate instances are a lot easier.

**Comment by mrovnanik** [comment]
OK, David, I've got the idea your. Thanks for your effort. Once again, I really appreciate what you are doing.

And the reason why I was attempting to put more Moquis on a single Tomcat server was the fact, that my knowledge on network issues is by far the least developed one. Not telling, the other are overdeveloped :-). I was happy having a single Tomcat running, being available on standard ports under my domain domain. I just do not know how to configure multiple Tomcats (using Docker, maybe) on a single host, so that they behave like independent servers (with their own names, e.g. srv06.rovnanik.sk, srv07.rovnanik.sk, etc. I did not want the customers to type weird ports (that is how I understood the Docker concept). There is probably a solution to that. I shall consult someone on that.

Finished for today, literally. Hope those guys in Turkey will settle things with minimum damage and as soon as possible. This work week is ending quite horribly.

**Comment by jonesde** [comment]
Sorry to hear your work week is going horribly...

Without using a VM or container you can run multiple Tomcat or Jetty instances on different ports with Apache in front of them (using mod_proxy or the like) to send requests for each domain to the proper instance. Moqui itself can handle multiple domains too with different root screens for each, but again in the same instance (I use this for the www.moqui.org server, which also hosts dejc.com and a couple other little sites).

With VMs (and maybe with containers, I have very close to zero experience with them) if you have multiple physical network interfaces you can send different IP addresses to different VMs, and maybe there are fancier ways too (i.e. support forwarding multiple external IPs through a single physical network port to different VMs).

On a side note, it is often best to use Apache and mod_proxy with Tomcat (was definitely that way years ago), though Jetty handles direct connections better with its threading and queuing model... even then many run Apache in front of it on the same or a different server to also serve up (or cache) static content, for SSL termination, and better handling of DoS attacks and other security issues.

**Comment by shendepu** [comment]
@mrovnanik you might look at https://github.com/shendepu/docker-moqui-mantle. It has the Dockerfile to build moqui-runtime docker image (the moqui-plus-runtime.war is built on local), and also a docker-compose.yml. 

If you want to run multiple tomcat containers on a single host, it is simple to to change line 14  (- 6000:8080) for each docker-compose.yml (tomcat container) with a different port (6000 to another port). 

The docker-entrypoint.sh file also adds RemoteIpValve to tomcat server.xml, so it is easy to add nginx with https in front moqui without no additional effort.

**Comment by mrovnanik** [comment]
Thank you guys. There is still a lot I can learn. That's what I get to know, everyday.

I will probably go for the first option David suggested, the Apache web server in front of Jetty (running Moqui) instances. Hope I can get the Apache running with SSL. It was a pain to get it started on the Tomcat.

If any of you has a documentation (installation notes, text tutorials, ...) that can minimize the effort needed to prepare such architecture, we can make a deal - I can pay for it.

@jonesde sorry for the confussion, by the horrible end of weak I meant all those things happening around the globe - the terror in Nice on Thursday, then the coup in Turkey on Friday. My work week was a bit better, although I can imagine myself be more productive.
